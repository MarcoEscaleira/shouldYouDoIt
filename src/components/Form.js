import React from 'react';
import PropTypes from 'prop-types';

const Form = (props) => (
  <div className="questionForm">
    <h1 className="questionForm__title">What do you want to do?</h1>
    <form 
      className="questionForm__form" 
      onSubmit={props.handleFormSubmit} 
      autoComplete="off"
    >
      <input 
        type="text" 
        name="toDo" 
        value={props.text} 
        onChange={props.handleTextInput}
        className="questionForm__form__input"
      />
      <button 
        type="submit" className="questionForm__form__btn"
      >
        Decide
      </button>
    </form>
  </div>
);

Form.propTypes = {
  handleFormSubmit: PropTypes.func,
  text: PropTypes.string,
  handleTextInput: PropTypes.func
}

export default Form;