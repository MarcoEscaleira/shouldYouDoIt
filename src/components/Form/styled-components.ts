import styled from "styled-components";
import { sizes, pallete } from "~/styles/_settings";

export const Container = styled.div`
  position: relative;
  text-align: center;
  display: flex;
  flex-direction: column;
  width: 30rem;

  @media (min-width: ${sizes.desktop_breakpoint}) {
    width: 50rem;
  }
`;

export const Title = styled.h1`
  color: ${pallete.sixth_light};
  font-size: 2.5rem;

  @media (min-width: ${sizes.desktop_breakpoint}) {
    font-size: 3rem;
  }
`;

export const FormContainer = styled.form`
  display: flex;
  justify-content: space-between;
  flex-direction: column;
`;

export const Input = styled.input`
  height: ${sizes.xl};
  flex-grow: 1;
  background-color: ${pallete.second_dark};
  border: none;
  border-radius: 1rem;
  color: ${pallete.white};
  font-size: 1.3rem;
  padding: ${sizes.s};
  margin-bottom: ${sizes.s};
  outline: none;
`;

export const Submit = styled.button`
  height: ${sizes.l};
  flex-grow: 1;
  background-color: ${pallete.fourth_dark};
  border: none;
  border-radius: 1rem;
  color: ${pallete.white};
  font-size: 1.4rem;
  font-weight: 300;
  letter-spacing: 0.05rem;
  margin-bottom: 1.5rem;
  outline: none;
`;
