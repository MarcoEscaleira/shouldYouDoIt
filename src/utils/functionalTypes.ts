export type NullaryFn<R> = () => R;
export type UnaryFn<A, R> = (a: A) => R;
export type BinaryFn<A, B, R> = (a: A, b: B) => R;
export type TernaryFn<A, B, C, R> = (a: A, b: B, c: C) => R;
export type QuadFn<A, B, C, D, R> = (a: A, b: B, c: C, d: D) => R;
export type QuinFn<A, B, C, D, F, R> = (a: A, b: B, c: C, d: D, f: F) => R;
export type SixFn<A, B, C, D, F, H, R> = (
  a: A,
  b: B,
  c: C,
  d: D,
  f: F,
  h: H
) => R;
