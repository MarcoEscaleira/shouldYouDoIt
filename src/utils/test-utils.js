import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import store from '../store';

export const renderWithRedux = (ui, initialState) => {
  return {
    ...render(
      <Provider store={store(initialState)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = ui => {
  return {
    ...render(<Router>{ui}</Router>)
  };
};

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
