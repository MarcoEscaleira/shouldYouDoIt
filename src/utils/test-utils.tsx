import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import store, { initialStoreState } from "../store";

export const renderWithRedux = (ui: JSX.Element, initialState?: {}) => {
  const state = {
    ...initialStoreState,
    ...initialState
  };
  return {
    ...render(
      <Provider store={store(state)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = (ui: JSX.Element) => ({
  ...render(<Router>{ui}</Router>)
});

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
