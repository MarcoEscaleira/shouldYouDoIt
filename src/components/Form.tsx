import React, { Fragment, useState, useCallback } from 'react';
import { connect } from 'react-redux';
import { actions, selectors } from '~store/decisions';
import { API_URL } from '../config';
import { UnaryFn, NullaryFn, BinaryFn } from '~utils/functionalTypes';
import ErrorModal from './ErrorModal';
import { StoreState } from '~store';

type SetDataType = BinaryFn<string, string, void>;
type ResetDataType = NullaryFn<void>;
type SetLoadingAnimationType = NullaryFn<void>;

type Props = {
  setData: SetDataType,
  resetData: ResetDataType,
  setLoadingAnimation: SetLoadingAnimationType,
  oldTextInput: string
};

const request = async (textInput: string, setData: SetDataType) => {
  const response = await fetch(`${API_URL}/?search=${textInput}`, {
    headers: {
      'Content-Type': 'application/json'
    }
  });
  const data = await response.json();

  setData(data, textInput);
};

const handleTextInput = (textInput: string, setTextInput: UnaryFn<string, void>) => {
  if (!textInput || /^[a-z\d\-_\s]+$/i.test(textInput)) {
    setTextInput(textInput.trim());
  }
};

const handleFormSubmit = (
  event,
  resetData: ResetDataType,
  textInput: string,
  oldTextInput: string,
  handleErrorModal: UnaryFn<string, void>,
  setLoadingAnimation: SetLoadingAnimationType,
  setData: SetDataType
) => {
  event.preventDefault();
  resetData();

  if (!textInput) {
    handleErrorModal('Please enter what you want to do');
    return;
  }

  if (oldTextInput === textInput) {
    handleErrorModal('Please do another thing!');
    return;
  }

  setLoadingAnimation();

  request(textInput, setData);
};

const Form: React.FC<Props> = ({ resetData, oldTextInput, setLoadingAnimation, setData }) => {
  const [textInput, setTextInput] = useState<string>('');
  const [showModal, setShowModal] = useState<boolean>(false);
  const [modalMessage, setModalMessage] = useState<string>('');

  const handleErrorModal = useCallback(() => {
    resetData();
    setShowModal(!showModal);
    setModalMessage(textInput || '');
  }, [showModal, textInput]);

  return (
    <Fragment>
      <div className="questionForm">
        <h1 className="questionForm__title">What do you want to do?</h1>
        <form
          className="questionForm__form"
          onSubmit={e => handleFormSubmit(e, resetData, textInput, oldTextInput, handleErrorModal, setLoadingAnimation, setData)}
          autoComplete="off"
        >
          <input
            type="text"
            name="toDo"
            value={textInput}
            onChange={e => handleTextInput(e.target.value, setTextInput)}
            className="questionForm__form__input"
            data-testid="input"
          />
          <button type="submit" className="questionForm__form__btn">
            Decide
          </button>
        </form>
      </div>
      <ErrorModal
        show={showModal}
        text={modalMessage}
        closeCallback={handleErrorModal}
      />
    </Fragment>
  );
};

const mapStateToProps = (state: StoreState) => ({
  oldTextInput: selectors.getOldTextInput(state)
});

const mapDispatchToProps = {
  resetData: actions.resetData,
  setLoadingAnimation: actions.setLoadingAnimation,
  setData: actions.setData
};

// @ts-ignore
export default connect(mapStateToProps, mapDispatchToProps)(Form);
