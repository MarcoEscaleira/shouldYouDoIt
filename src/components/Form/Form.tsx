import React, { Fragment, useState } from "react";
import { connect } from "react-redux";
import { UnaryFn, NullaryFn, BinaryFn } from "../../utils/functionalTypes";
import { actions as decisionsActions, selectors } from "../../store/decisions";
import { actions as errorModalActions } from "../../store/errorModal";
import { API_URL } from "../../config";
import { StoreState } from "../../store";

import {
  Container,
  Title,
  FormContainer,
  Input,
  Submit
} from "./styled-components";

type SetDataType = BinaryFn<string, string, void>;
type ResetDataType = NullaryFn<void>;
type SetLoadingAnimationType = NullaryFn<void>;

type Props = {
  setData: SetDataType;
  resetData: ResetDataType;
  setLoadingAnimation: SetLoadingAnimationType;
  oldTextInput: string;
  openErrorModal: UnaryFn<string, void>;
};

const request = async (textInput: string, setData: SetDataType) => {
  const response = await fetch(`${API_URL}/?search=${textInput}`, {
    headers: {
      "Content-Type": "application/json"
    }
  });
  const data = await response.json();

  setData(data, textInput);
};

const handleTextInput = (
  textInput: string,
  setTextInput: UnaryFn<string, void>
) => {
  if (!textInput || /^[a-z\d\-_\s]+$/i.test(textInput)) {
    setTextInput(textInput.trim());
  }
};

const handleFormSubmit = (
  event,
  resetData: ResetDataType,
  textInput: string,
  oldTextInput: string,
  setLoadingAnimation: SetLoadingAnimationType,
  openErrorModal: UnaryFn<string, void>,
  setData: SetDataType
) => {
  event.preventDefault();
  resetData();

  if (!textInput) {
    openErrorModal("Please enter what you want to do.");
    return;
  }

  if (oldTextInput === textInput) {
    openErrorModal("Please do another thing!");
    return;
  }

  setLoadingAnimation();

  request(textInput, setData);
};

const Form: React.FC<Props> = ({
  resetData,
  oldTextInput,
  setLoadingAnimation,
  openErrorModal,
  setData
}) => {
  const [textInput, setTextInput] = useState<string>("");

  return (
    <Fragment>
      <Container>
        <Title>What do you want to do?</Title>
        <FormContainer
          onSubmit={e =>
            handleFormSubmit(
              e,
              resetData,
              textInput,
              oldTextInput,
              setLoadingAnimation,
              openErrorModal,
              setData
            )
          }
          autoComplete="off"
        >
          <Input
            type="text"
            name="toDo"
            value={textInput}
            onChange={e => handleTextInput(e.target.value, setTextInput)}
            data-testid="input"
          />
          <Submit type="submit">Decide</Submit>
        </FormContainer>
      </Container>
    </Fragment>
  );
};

const mapStateToProps = (state: StoreState) => ({
  oldTextInput: selectors.getOldTextInput(state)
});

const mapDispatchToProps = {
  resetData: decisionsActions.resetData,
  setLoadingAnimation: decisionsActions.setLoadingAnimation,
  setData: decisionsActions.setData,
  openErrorModal: errorModalActions.openModal
};

// @ts-ignore
export default connect(mapStateToProps, mapDispatchToProps)(Form);
