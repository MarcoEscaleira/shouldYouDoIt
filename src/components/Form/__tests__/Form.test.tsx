import React from "react";
import Form from "../Form";
import { renderWithRedux } from "~/utils/test-utils";

describe("Form test suite", () => {
  it("Should render Form correctly", () => {
    const { container } = renderWithRedux(<Form />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
