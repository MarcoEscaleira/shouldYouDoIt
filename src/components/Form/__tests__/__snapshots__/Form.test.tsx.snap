// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form test suite Should render Form correctly 1`] = `
<div
  class="sc-bdVaJa dKdJKp"
>
  <h1
    class="sc-bwzfXH jjfDCS"
  >
    What do you want to do??
  </h1>
  <form
    autocomplete="off"
    class="sc-htpNat bdTqtb"
  >
    <input
      class="sc-bxivhb cNpncz"
      data-testid="input"
      name="toDo"
      type="text"
      value=""
    />
    <button
      class="sc-ifAKCX bDTMJK"
      type="submit"
    >
      Decide
    </button>
  </form>
</div>
`;
